torch/csrc/jit/pybind_utils.h

```diff
@@ -170,6 +170,25 @@ inline InferredType tryToInferContainerType(py::handle input) {
         return type_match.reason();
       }
     }
+    if (py::hasattr(input, "_fields")) {
+      // If the tuple is a NamedTuple type
+      auto qualifiedName = c10::QualifiedName(py::cast<std::string>(
+          py::module::import("torch.jit").attr("_qualified_name")(input.get_type())));
+      auto PyCu = get_python_cu();
+      c10::NamedTypePtr named_tuple_type = PyCu->get_named_tuple(qualifiedName);
+      if (named_tuple_type != nullptr) {
+        // If the python_cu already contains the type, just return it
+        return InferredType(named_tuple_type);
+      } else {
+        // Otherwise fetch the names correspondingly and create a new NamedTuple type
```

Comment:
This code is a reimplementation of the logic in PythonResolver::resolveType. They need to unified into a single function can called from both location. For instance the check for _fields here is just a reimplementation of isNamedTupleClass in python resolver.

Reply:
I actually tried merging these two together, there're some problems: 1. We don't expose SourceRange to `tryToInferType` so we couldn't throw ErrorReport like resolveType. (Maybe we can get the python interpreter SourceRange here) 2. we could not call get_named_tuple_property function here because it will try to resolve the python type annotation again (with `ann_to_type`) where it does not exist in tracing (we need to pass in `rcb` to `ann_to_type` to know lazily created types from python annotations).